[24] event work manager

steve1rm opened the pull request:
# Pull Request template

###  What does this PR solve?

- Working Sending the multipart request for events using work manager

Commit:
Adding work manager dependencies and service endpoints for creating, updating, deleting, events

Commit:
Adds the multipart parameters to the event service

Commit:
Adds MutlipartBody Part for the photos that have been converted to bytearray

Commit:
Adds injection for UploadEventWorker

Commit:
Adds worker injection to BusbyTaskyApplication and androidManifest

Commit:
Has error trying to find the path of the selected image

Commit:
Can send event request using multipart but with sending empty photos. Exception with photos multipart

Commit:
Made comment about work manager

Commit:
Made comment about work manager

philipplackner commented on data/src/main/java/me/androidbox/data/local/converter/EventConverter.kt:
```diff
@@ -0,0 +1,25 @@
+import me.androidbox.data.remote.model.request.EventRequestDto
+import javax.inject.Inject
+
+class EventConverter @Inject constructor(){
+    private val moshi = Moshi.Builder()
+        .add(KotlinJsonAdapterFactory())
+        .build()
+
+    private val jsonAdapter = moshi.adapter(EventRequestDto::class.java)
+
+    @TypeConverter
+    fun fromJson(json: String): EventRequestDto? {
+        return jsonAdapter.fromJson(json)
+    }
+
+    @TypeConverter
+    fun toJson(eventRequestDto: EventRequestDto): String {
+        return jsonAdapter.toJson(eventRequestDto)
+    }
+}
```
TypeConverters are only used for Room. You don't need this to manually serialize an object for a network request

steve1rm replied:
@philipplackner So I should be doing something like the following:
```
        /* Serialize the event request to json */
        val moshi = Moshi.Builder().build()
        val jsonAdapter = moshi.adapter<EventRequestDto>(EventRequestDto::class.java)
        val eventRequestJson = jsonAdapter.toJson(eventRequest)
```

philipplackner replied:
Yes 👍

philipplackner commented on data/src/test/java/me/androidbox/data/local/converter/EventConverterTest.kt:
```diff
@@ -0,0 +1,53 @@
+import java.util.*
+import kotlin.random.Random
+
+class EventConverterTest {
+
+    private val eventConverter = EventConverter()
+
+    @Test
+    fun `should serialize the event request into json string`() {
+        // Arrange
+        val eventRequestDto = generateEventRequestDto()
+
+        // Act
+        val actual = eventConverter.toJson(eventRequestDto)
+
+        // Assert
+        Truth.assertThat(actual).isNotEmpty()
+    }
+
+    @Test
+    fun `should deserialize json string into a event result object`() {
+        // Arrange
+        val jsonString = "{\"id\":\"3061eb15-154b-4ea8-b89a-984cf16fa6d6\",\"title\":\"Title\",\"description\":\"description\",\"from\":3,\"to\":6,\"remindAt\":4,\"attendeeIds\":[\"25921c52-3e59-4acb-9c81-f79c8fbc2a27\",\"cd30b0a2-4015-43cb-b45d-70535bc31bb5\",\"8d1ed082-5859-4461-a1e6-268ff96d5ab9\",\"32ffd738-6a5f-4938-8f75-59302bd0443d\"]}"
+
+        // Act
+        val actual = eventConverter.fromJson(jsonString)
+
+        // Assert
+        Truth.assertThat(actual).isNotNull()
+        Truth.assertThat(actual?.id).isEqualTo("3061eb15-154b-4ea8-b89a-984cf16fa6d6")
+    }
+}
```
What you're testing here is nothing else than the Moshi library. You can be sure that the authors of that library already considered testing. You never really test other SDKs, but rather just your own code.

steve1rm replied:
@philipplackner Removed the unit tests

philipplackner commented on presentation/src/main/java/me/androidbox/presentation/HomeActivity.kt:
```diff
@@ -21,13 +26,39 @@ class HomeActivity : ComponentActivity() {
+                }
+            }
+
+            val eventState by derivedStateOf {
+                val downloadEvent = downloadInfo?.outputData?.getString("")?.toUri()
+            }
```
This `derivedStateOf` does nothing, since you don't return any state in the block (and if you do, it would need to be wrapped with `mutableStateOf()`)

steve1rm replied:
@philipplackner I was just trying this out and forgot to remove it.

philipplackner commented on presentation/src/main/java/me/androidbox/presentation/util/imp/CreatePhotoMultipartImp.kt:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private suspend fun getByteArrayFromUri(uri: Uri): ByteArray {
+       return withContext(Dispatchers.Default) {
```
Shouldn't this use the `IO` dispatcher since reading bytes is an IO operation?

steve1rm replied:
@philipplackner Changed to the IO dispatcher

steve1rm replied:
@philipplackner This is the part I am stuck on. Does this code look correct?

philipplackner commented on presentation/src/main/java/me/androidbox/presentation/util/imp/CreatePhotoMultipartImp.kt:
```diff
@@ -0,0 +1,65 @@
+content://com.android.providers.media.documents/document/image%3A38337
+* */
+
+class CreatePhotoMultipartImp @Inject constructor(@ApplicationContext private val context: Context) : CreatePhotoMultipart {
```
What do you gain by having an abstraction for this?

philipplackner commented on presentation/src/main/java/me/androidbox/presentation/worker/UploadEventWorker.kt:
```diff
@@ -0,0 +1,105 @@
+package me.androidbox.presentation.worker
```
WorkManager workers are rather data related, especially if they upload something like here

steve1rm replied:
@philipplackner so this should belong in my Data Layer module. I did have it there ordinary but moved it to the presentation module because that is where the work would be triggered from.

philipplackner replied:
No, the work could as well be triggered from your data layer in the repository. That's where the request is executed, the worker is just a tool to execute it reliably

MKiperszmid replied:
A way to trigger it from Presentation, would be to abstract it in the Domain layer, and have the implementation on the data layer

Kinda like what you did for the Email Validator

philipplackner commented on presentation/src/main/java/me/androidbox/presentation/worker/UploadEventWorker.kt:
```diff
@@ -0,0 +1,105 @@
+
+        /* These are just a random set of photos that I got from selecting photos from the device.
+         * I just copied them here to try and send them as a multi-part request just to test if I can send them */
+        val listOfPhotos = listOf(
+            "content://com.google.android.apps.photos.contentprovider/0/1/mediakey%3A%2Flocal%253A44f502c3-7f36-4f9b-887c-5d8f2f45f3cc/ORIGINAL/NONE/image%2Fjpeg/430889281",
+            "content://com.android.providers.media.documents/document/image%3A38324",
+            "content://com.google.android.apps.photos.contentprovider/-1/1/content%3A%2F%2Fmedia%2Fexternal%2Fimages%2Fmedia%2F37452/ORIGINAL/NONE/image%2Fjpeg/640889962")
```
AFAIK, these may differ each time you retrieve them. Not tested, but have heard people say that. I'd test this with a Uri you really picked on the device (this could explain your `FileNotFoundException`).

Commit:
Removed code from HomeActivity

Commit:
Using json adapter to serialize

Commit:
Changed to IO dispatcher

Commit:
Issue with getting the moshi jsonAdapter working

Commit:
Merge branch 'develop' of github.com:steve1rm/BusbyTasky into 24-event-work-manager

Commit:
Adds media photo path

Commit:
Updates comments

Commit:
Removed comments

Commit:
Merge branch 'develop' of github.com:steve1rm/BusbyTasky into 24-event-work-manager

Commit:
Merge branch 'develop' into 24-event-work-manager

Commit:
Merge branch 'develop' of github.com:steve1rm/BusbyTasky into 24-event-work-manager

Commit:
1) Can login and register new account success
2) Cleaned up the HomeActivity
3) Send a request via work manager with multi-part request that contains photos
4) Work manager fails with some error even those the request was 200 OK

Commit:
Removed comments

Commit:
Removed comments

Commit:
Removed comment and remove staring a servide to tigger the notication

Commit:
Fixes issue eventCreatorId had to be replaced with host

steve1rm commented on presentation/src/main/java/me/androidbox/presentation/worker/UploadEventWorker.kt:
```diff
@@ -0,0 +1,98 @@
+        val result = responseResult.fold(
+            onSuccess = { eventDto ->
+                val outputData = workDataOf("eventKey" to eventDto)
+                Result.success(outputData)
```
@MKiperszmid @philipplackner Is this correct on how to pass data back from the work manager: I am getting this crash:
```
Work [ id=eb508899-6586-4674-94f7-8a17420983f0, tags={ me.androidbox.presentation.worker.UploadEventWorker } ] failed because it threw an exception/error
    java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Key eventKeyhas invalid type class me.androidbox.data.remote.model.response.EventDto
```

steve1rm replied:
Another question, what do we do with this EventDto object that is returned from the endpoint? I save the event into Room DB before I sending the request via work manager.

philipplackner replied:
Yes it's correct, but you can only pass primitive types (and maybe Parcelables I guess). It just doesn't know how to pass your `EventDto` back as a result. You can try to make it a `Parcelable` and if that doesn't work pass a serialized JSON string of the DTO object.
Either way, I wonder why you need to pass the whole event back? Why not just the info that it was successful?

Commit:
removed unused import

Commit:
Merge branch 'develop' of github.com:steve1rm/BusbyTasky into 24-event-work-manager

Commit:
Adds dependency for creating a workmanager and uploadEventWorkManager

Commit:
Adds uploadEvent interface and implementation

Commit:
WIP work manager getting the data and mapping the event to dto objects

Commit:
Removed commented code

Commit:
Merge branch 'develop' of github.com:steve1rm/BusbyTasky into 24-event-work-manager

Commit:
Merge branch 'develop' of github.com:steve1rm/BusbyTasky into 24-event-work-manager

Commit:
Updated the mapper to create the updateEvent and createEvent requests

Commit:
Can serialize the json for the create and update event using moshi

Commit:
Adds compiler for the workmanager can be provided. Cleanup UploadEventImp

Commit:
Completed the work manager for UploadEventWorker

Commit:
Merge branch 'develop' of github.com:steve1rm/BusbyTasky into 24-event-work-manager

Commit:
Inlined responseState when insertInserting a new event in the EventViewModel

steve1rm commented on data/src/main/java/me/androidbox/data/worker_manager/UploadEventImp.kt:
```diff
@@ -0,0 +1,58 @@
+            workDataOf(
+                IS_EDIT_MODE to isEditMode,
+                EVENT to eventRequestJson)
+        }
```
@philipplackner @MKiperszmid Not sure if this is worth doing. I am just thinking its not worth adding the photos if the user didn't select any photos.

MKiperszmid replied:
Tbh this is extra code that doesn't really do much
If you send an empty photo list, backend will not do anything 
So it can be removed

steve1rm commented on data/src/main/java/me/androidbox/data/worker_manager/UploadEventWorker.kt:
```diff
@@ -0,0 +1,100 @@
+                        val outputData = photoToJson(eventDto)
+
+                        Result.success(outputData)
+                    }
```
@philipplackner @MKiperszmid Will return the photos url and keys in the Success, if there any photos that was uploaded when creating or updating an event. Is this correct?

MKiperszmid replied:
Yea but the question is, do you really care about the response of the Success?

steve1rm replied:
@MKiperszmid I am sure that @philipplackner said we get back the key and url in the response. And these are the keys we use when we want to delete the photos from the BE. Maybe @philipplackner can confirm what we need to do with this response.

philipplackner replied:
Yes correct @steve1rm . You need these keys and the URLs are also only known after the API responds

steve1rm commented on data/src/main/java/me/androidbox/data/worker_manager/UploadEventImp.kt:
```diff
@@ -0,0 +1,58 @@
+            .setInputData(inputData = eventInputData).build()
+
+        workManager.enqueue(uploadWorkerRequest)
+    }
```
@MKiperszmid @MKiperszmid What is the best way to handle if there is no internet?
i.e. The user has lost connection, this will fail. Would the user have to click save again to upload the event if the first time it fails.

What happens if the there is no internet connection so the upload will fail. But if the connection is suddenly restored what will be the best way to handle that case?

MKiperszmid replied:
If i'm not mistaken, WorkManager should automatically handle it

Adding a backoff (previous comment) will make it so once there's connection, it will retry again

Tho we can wait for @philipplackner to fully confirm this

philipplackner replied:
When using the network connected constraints for WorkManager, that's handled. Ideally together with a backoff strategy

steve1rm commented on data/src/main/java/me/androidbox/data/worker_manager/UploadEventWorker.kt:
```diff
@@ -0,0 +1,100 @@
+            val listOfPhotoMultiPart = listOfPhoto?.let {
+                /* Create the multipart for the list of photos */
+                createPhotoMultipart.createMultipartPhotos(listOfPhoto)
+            } ?: listOf()
```
@philipplackner @MKiperszmid when setting the parameters in the `UpLoadEventImp` I added a check to only add the photos to the inputData if the photo list wasn't empty. So I guess having this check here would mean its not worth checking as the inputData would be null if there wasn't any added photos in the first place.

MKiperszmid replied:
I would remove the other stuff and just have this here

steve1rm replied:
@MKiperszmid Thanks for confirming this

MKiperszmid commented on data/src/main/java/me/androidbox/data/remote/network/EndPointConstant.kt:
```diff
@@ -5,4 +5,5 @@ object EndPointConstant {
     const val LOGIN = "login"
     const val LOGOUT = "logout"
     const val AUTHENTICATE =  "authenticate"
+    const val EVENT = "event"
```
You could just have a constant on each of the service
There's no added benefit on having an extra class

It's not like the constants will be shared between 2 or more classes

philipplackner replied:
There isn't even a real use case for having constants for URLs at all. They're just used at one place and a constant also doesn't add readability

MKiperszmid commented on data/src/main/java/me/androidbox/data/worker_manager/UploadEventImp.kt:
```diff
@@ -0,0 +1,58 @@
+                EVENT to eventRequestJson)
+        }
+
+        val uploadWorkerRequest  = OneTimeWorkRequestBuilder<UploadEventWorker>()
```
What if the upload fails? You could add some backoff time 

Check "setBackoffCriteria"

MKiperszmid commented on data/src/main/java/me/androidbox/data/worker_manager/UploadEventWorker.kt:
```diff
@@ -0,0 +1,100 @@
+                    }
+                },
+                onFailure = { throwable ->
+                    val errorDate = workDataOf(ERROR to throwable)
```
Why is it named
errorDate?

steve1rm replied:
@MKiperszmid I named it errorData as I will create a workDataOf to show the error if the upload failed. What do you think this should be called?

philipplackner replied:
Yes, but there's a typo, you called it `errorDate`

steve1rm replied:
Should be errorData

MKiperszmid commented on data/src/main/java/me/androidbox/data/worker_manager/util/CreatePhotoMultipart.kt:
```diff
@@ -0,0 +1,7 @@
+
+import okhttp3.MultipartBody
+
+interface CreatePhotoMultipart {
```
Do you really need this interface?

philipplackner replied:
Also it rather sounds like a function and not an interface

MKiperszmid commented on presentation/src/main/java/me/androidbox/presentation/event/viewmodel/EventViewModel.kt:
```diff
@@ -146,17 +150,28 @@ class EventViewModel @Inject constructor(
-
-            val alarmItem = event.toAlarmItem(AgendaType.EVENT)
-            alarmScheduler.scheduleAlarmReminder(alarmItem)
+            when(val responseState = eventRepository.insertEvent(event)) {
```
In what cases will this ever return something different than success?
Since you are just inserting to a db

steve1rm replied:
@MKiperszmid I can remove it if there is nothing that can go wrong with an insert. @philipplackner

philipplackner replied:
The only thing I can think of is if the device is out of storage

Commit:
Removed checking for empty list as if the list is empty it will still send an empty list anyway. Uses .toTypedArray

philipplackner commented on data/src/main/java/me/androidbox/data/local/converter/EventConverter.kt:
```diff
@@ -0,0 +1,25 @@
+import me.androidbox.data.remote.model.request.EventCreateRequestDto
+import javax.inject.Inject
+
+class EventConverter @Inject constructor(){
```
If you have an @Inject constructor, you don't need a @Provides function for this class

Commit:
Adds backOffCritieria

Commit:
Adds runAttemptCount and retry

steve1rm commented on data/src/main/java/me/androidbox/data/worker_manager/UploadEventImp.kt:
```diff
@@ -0,0 +1,55 @@
+            .setBackoffCriteria(
+                BackoffPolicy.EXPONENTIAL,
+                WorkRequest.MIN_BACKOFF_MILLIS,
+                TimeUnit.SECONDS)
```
@philipplackner  @MKiperszmid Is this the best way to handle a backoffcriteria?

philipplackner replied:
Yes, just that you assign a milliseconds value but specify it as seconds

steve1rm commented on data/src/main/java/me/androidbox/data/worker_manager/UploadEventWorker.kt:
```diff
@@ -0,0 +1,106 @@
+
+            if(runAttemptCount > RETRY_COUNT) {
+                Result.failure()
+            }
```
@philipplackner @MKiperszmid Adds a RETRY_COUNT

steve1rm commented on data/src/main/java/me/androidbox/data/worker_manager/UploadEventWorker.kt:
```diff
@@ -0,0 +1,106 @@
+                },
+                onFailure = { _ ->
+                    Result.retry()
+                }
```
@philipplackner @MKiperszmid Adds retry in failure.

Commit:
Changed timeunits to milli seconds
